### DISCUSSION TITLE
generating spherical surface mesh from points on surface of sphere

### DISCUSSION THREAD

**Discussion by tom-new** [pyvista/pyvista#5804] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
import pyvista as pv

cell = [4,0,1,2,0]
points = [[1.,0.,0.],[0.,1.,0.],[0.,0.,1.]]
cell_type = np.array([4]) # poly line
polygon = pv.UnstructuredGrid(cell, cell_type, points)

sphere = pv.Icosphere(nsub=6)

surface = polygon.merge(sphere)

surface.plot()
```

### Description

I have a number of points on the surface of the sphere which form a [spherical polygon](https://en.wikipedia.org/wiki/Spherical_trigonometry#Spherical_polygons). I would like to generate a mesh that corresponds the portion of the sphere that the spherical polygon covers. As you can see above, I attempted this by merging a mesh created from the spherical polygon with a spherical surface mesh, but it does not give me the +x/+y/+z octant of the sphere as I expected (instead it gives just full spherical surface).

Is there a straighforward way to achieve this using PyVista's built-in functions that I am missing? If a more custom-made solution is necessary, I would be immensely grateful for any pointers in the right direction for how to approach this problem.

### System Information

```shell
Date: Thu Mar 21 15:33:34 2024 AEDT

                OS : Darwin
            CPU(s) : 10
           Machine : arm64
      Architecture : 64bit
               RAM : 64.0 GiB
       Environment : IPython
       File system : apfs
        GPU Vendor : Apple
      GPU Renderer : Apple M1 Max
       GPU Version : 4.1 Metal - 88
  MathText Support : False

  Python 3.12.0 (main, Nov 14 2023, 12:57:43) [Clang 15.0.0
  (clang-1500.0.40.1)]

           pyvista : 0.42.3
               vtk : 9.3.0rc2
             numpy : 1.26.2
        matplotlib : 3.8.1
            scooby : 0.9.2
             pooch : 1.8.0
            pillow : 10.1.0
           IPython : 8.17.2
        ipywidgets : 8.1.1
             scipy : 1.11.3
        jupyterlab : 4.0.8
      nest_asyncio : 1.5.8
```

### Screenshots

![Screenshot 2024-03-21 at 15 34 20](https://github.com/pyvista/pyvista/assets/46676673/d0c222f1-1b29-46ef-b6a3-de28a90b2be8)

**Comment by MatthewFlamm** [comment]
`merge` is additive, which results in the behavior you observe.  Your `polygon` is actually a triangle that is not coincident on the sphere surface.  What is your end goal?  If you are looking for an octant of a sphere.  This can be done with the `theta` and `phi` parameters to `Sphere`.

**Reply by tom-new** [reply]
The points are coincident to the sphere and form a spherical polygon, and I would like to generate a surface mesh which corresponds to that spherical polygon. The fact that the surface of _this_ spherical polygon is an octant is not important, I just chose a simple example.

**Comment by darikg** [comment]
I think you're looking for some sort of clipping procedure. The [clipping with a surface](https://docs.pyvista.org/version/stable/examples/01-filter/clipping-with-surface.html) shows one way to do that. ~~I think that would require you to manually construct a clipping "ribbon" mesh perpendicular to the sphere surface, centerd on the great arcs of the polygon boundary, which sounds doable but annoying.~~ (Edit on second thought a really lazy and easy way to do it is just repeatedly clip with a plane passing through every great circle belonging to the polygon, keeping the portion of the mesh in the half-space consistent with the polygon interior.)

Alternatively there's some more powerful features in [vtkClipPolyData](https://vtk.org/doc/nightly/html/classvtkClipPolyData.html) that I don't think are fully exposed in pyvista yet, but it's still pretty easy to use the VTK python bindings to access them. [This C++ example](https://examples.vtk.org/site/Cxx/PolyData/ImplicitSelectionLoop/) illustrates what I'm talking about. In that particular example they use `vtkImplicitSelectionLoop` to calculate the distance from a finite set of points on the surface to do the clipping; which in your case would require subsampling the the spherical polygon along its boundary until you get smooth enough clipping. 

Finally, I think it's possible to pass a completely custom implicit function to vtkClipPolyData. Is there a reasonably easy way to construct a signed distance function for a spherical polygon? I.e. a function that takes a latitude and longitude and outputs a negative distance for points inside the polygon, a positive distance for points outside, and zero on the boundary? I think that would do it.

**Reply by darikg** [reply]
going back to the lazy easy idea:

```python
from pyvista import Sphere, Plotter, Plane
from numpy import array, roll, cross

sphere = Sphere(radius=1)
pts = array([[1.,0.,0.], [0.,1.,0.] ,[0.,0.,1.]])
mesh = sphere.copy()

# For each segment in the boundary
for a, b in zip(pts, roll(pts, -1, axis=0)):  
    mesh = mesh.clip(normal=cross(b, a))

pl = Plotter()
pl.add_mesh(mesh)
for pt in pts:
    pl.add_mesh(Sphere(center=pt, radius=0.1), color='red')
pl.show()
```

![image](https://github.com/pyvista/pyvista/assets/6875882/834ff1b7-812a-4ca8-a16a-9faff4dfd605)

**Reply by darikg** [reply]
Of course this will only work if your polygons are convex, otherwise the more complicated stuff is required

**Reply by MatthewFlamm** [reply]
Another option is to start with a meshed polygon and then project the points onto the sphere surface.

**Reply by darikg** [reply]
Here's the `clip_surface` approach, which should work with convex polygons. Constructing the clipper was easier than I thought'd it be. Spherical geometry is really convenient.

```python
from pyvista import Sphere, Plotter, Plane, PolyData
from numpy import array, roll, concatenate, arange, stack, append, ones

sphere = Sphere(radius=1).subdivide(3)
pts = array([[1.,0.,0.], [0.,1.,0.] ,[0.,0.,1.]])

# To be outside the sphere needs to be scaled by sqrt(2)
# Use 1.5 for safety margin
scaled_pts_and_origin = concatenate([1.5 * pts, [[0, 0, 0]]])

# for pts = [a, b, c, d, ...]
# polyon edges (a, b), (b, c), (c, d), ...
# construct one triangle per edge with the origin
#  So triangles (a, b, origin), (b, c, origin), ...
n = len(pts)
faces = stack([
    arange(n),
    roll(arange(n), -1),
    n * ones(n, dtype='int')
], axis=1) # (n_pts * 3)

clipper = PolyData.from_regular_faces(scaled_pts_and_origin, faces)
clipped = sphere.clip_surface(clipper)
clipped.plot()
```

**Reply by tom-new** [reply]
Thanks both very much for the suggestions. Now that I know what it's called, it's obvious that clipping is the functionality I was looking for, but it didn't catch my attention when I was looking through the docs.

In general however, I need to be able to do this for an arbitrary spherical polygon—they may be concave or self-intersecting, or even have interior rings which should have "no mesh" in _their_ interior (though the latter would only be a nice-to-have). Creating an SDF for a spherical polygon seems straightforward at first glance, but if the polygon contains either pole or part of it crosses the antimeridian then I would expect an SDF created for 2D Cartesian coordinates to break when applied to lat-lon coordinates.

Perhaps this indicates I should look into Matthew's suggestion of projecting a flat mesh to the surface. It was also something I tried before I posted this question, however I just couldn't wrap my head around how to use built-in methods to achieve this (I tried using `extrude`, but looking at it again it doesn't seem like it will achieve this).
